esrally/log.py

```diff
@@ -43,6 +44,44 @@ def log_config_path():
     return os.path.join(paths.rally_confdir(), "logging.json")
 
 
+def add_missing_loggers_to_config():
+    """
+    Ensures that any missing top level loggers in resources/logging.json are
+    appended to an existing log configuration
+    """
+    logger = logging.getLogger(__name__)
+
+    def missing_keys(source, target):
+        """
+        Returns any top-level dicts present in 'source', but not in 'target'
+        :return: A dict of all keys present in 'source', but not in 'target'
+        """
+        missing_keys = {}
+        for k in source:
+            if k in source and k in target:
+                continue
+            else:
+                missing_keys[k] = source[k]
+        return missing_keys
+
+    source_path = io.normalize_path(os.path.join(os.path.dirname(__file__), "resources", "logging.json"))
+    with open(log_config_path(), "r+", encoding="UTF-8") as target:
+        with open(source_path, "r+", encoding="UTF-8") as src:
+            template = json.load(src)
+            existing_logging_config = json.load(target)
+            existing_logging_config_copy = copy.deepcopy(existing_logging_config)
+            if missing_loggers := missing_keys(source=template["loggers"], target=existing_logging_config_copy["loggers"]):
+                logger.info(
+                    "Found loggers [%s] in source template that weren't present in the existing configuration, adding them.",
+                    str(missing_loggers),
+                )
+                existing_logging_config["loggers"].update(missing_loggers)
+                updated_config = json.dumps(existing_logging_config, indent=2)
+                target.seek(0)
+                target.write(updated_config)
+                target.truncate()
```

Comment:
I would suggest opening the target once in read mode, and then reopen it in write mode if there are missing loggers, which would avoid this tricky code.

Reply:
Addressed in https://github.com/elastic/rally/pull/1669/commits/cb52f45a85855619ff7f290a8f8a71ffd5d3559d